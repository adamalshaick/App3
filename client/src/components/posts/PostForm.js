import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import TextAreaFieldGroup from "../common/TextAreaFieldGroup";
import { addPost } from "../../actions/postActions";
import handleInputErrors from "../common/hoc/handleInputErrors";

class PostForm extends Component {
  constructor(props) {
    super(props);
    this.state = { text: "" };
  }

  onSubmit = e => {
    e.preventDefault();
    const { currentProfile } = this.props;
    const newPost = {
      text: this.state.text,
      name: currentProfile.user.name,
      avatar: currentProfile.profileImage
    };
    this.props.addPost(newPost);
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };
  render() {
    const { errors } = this.props;
    return (
      <div className="mb-3">
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <TextAreaFieldGroup
              className="form-control form-control-lg"
              placeholder="Add a post"
              name="text"
              value={this.state.text}
              onChange={this.onChange}
              error={errors.text}
            />
          </div>
          <button
            type="submit"
            className="btn btn-outline-primary btn-sm float-right"
          >
            Submit
          </button>
        </form>
      </div>
    );
  }
}

PostForm.propTypes = {
  addPost: PropTypes.func.isRequired,
  errors: PropTypes.object.isRequired,
  currentProfile: PropTypes.object.isRequired
};

export default connect(
  null,
  { addPost }
)(handleInputErrors(PostForm));
