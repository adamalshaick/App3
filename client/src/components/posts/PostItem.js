import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { deletePost, addLike, removeLike } from "../../actions/postActions";
import CommentForm from "../comments/CommentForm";
import CommentFeed from "../comments/CommentFeed";

export class PostItem extends Component {
  constructor(props) {
    super(props);
    this.state = { displayCommentForm: false };
  }
  onDeleteClick(id) {
    this.props.deletePost(id);
  }

  onLikeClick(id) {
    this.props.addLike(id);
  }

  onUnlikeClick(id) {
    this.props.removeLike(id);
  }

  render() {
    const { post, currentProfile } = this.props;

    return (
      <div className="row">
        <div className="row mt-3">
          <img
            style={{ width: "40px", height: "40px" }}
            className="rounded-circle  ml-4"
            src={post.avatar}
            alt=""
          />
          <div className="mt-1 ml-2">{post.name}</div>
        </div>
        <div className="col-12">
          <div style={{ fontSize: "0.9rem" }} className="float-left ml-2">
            {post.text}
          </div>
        </div>

        <div className="mt-2 ml-3 col-12">
          <div className="row">
            <div className="col-12">
              <button
                type="button"
                onClick={() => {
                  this.setState(prevState => ({
                    displayCommentForm: !prevState.displayCommentForm
                  }));
                }}
                className="btn btn-outline-primary btn-sm float-right"
              >
                Add a comment
              </button>

              {post.user === currentProfile.user._id ? (
                <button
                  onClick={this.onDeleteClick.bind(this, post._id)}
                  type="button"
                  className="btn btn-outline-danger btn-sm mr-1 float-right"
                >
                  Delete
                </button>
              ) : null}

              {post.likes.find(
                like => like.user === currentProfile.user._id
              ) ? (
                <button
                  onClick={this.onUnlikeClick.bind(this, post._id)}
                  type="button"
                  className="btn  btn-outline-danger btn-sm mr-1 float-right"
                >
                  <i className="fas fa-thumbs-down" />
                </button>
              ) : null}

              <button
                onClick={this.onLikeClick.bind(this, post._id)}
                type="button"
                className="btn btn-outline-success btn-sm mr-1 float-right"
              >
                <i className="fas fa-thumbs-up mr-2" />
                <small>{post.likes.length}</small>
              </button>
            </div>
          </div>
          <div className="mt-2 ml-lg-4 row">
            <div className="col-12">
              <CommentFeed postId={post._id} comments={post.comments} />
              {this.state.displayCommentForm ? (
                <CommentForm
                  postId={post._id}
                  currentProfile={currentProfile}
                />
              ) : null}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

PostItem.propTypes = {
  deletePost: PropTypes.func.isRequired,
  addLike: PropTypes.func.isRequired,
  removeLike: PropTypes.func.isRequired,
  post: PropTypes.object.isRequired,
  auth: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { deletePost, addLike, removeLike }
)(PostItem);
