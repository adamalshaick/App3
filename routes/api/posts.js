const express = require("express");
const router = express.Router();
const mongoose = require("mongoose");
const passport = require("passport");

// Post model
const Post = require("../../models/Post");
// Profile model
const Profile = require("../../models/Profile");

// Controllers
const postsController = require("../../controllers/postsController");

// Validation
const validatePostInput = require("../../validation/post");

// @route GET api/posts
// @desc Get post
// @access Public
router.get("/", (req, res) => {
  postsController.getPosts(req, res);
});

// @route GET api/posts/:id
// @desc Get post by id
// @access Public
router.get("/:id", (req, res) => {
  postsController.getPost(req, res);
});

// @route POST api/posts
// @desc Create post
// @access Private
router.post(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    postsController.createPost(req, res);
  }
);

// @route DELETE api/posts/:id
// @desc Delete post
// @access Private
router.delete(
  "/:id",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    postsController.deletePost(req, res);
  }
);

// @route POST api/posts/like/:id
// @desc Like post
// @access Private
router.post(
  "/like/:id",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    postsController.likePost(req, res);
  }
);

// @route POST api/posts/unlike/:id
// @desc Unlike post
// @access Private
router.post(
  "/unlike/:id",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    postsController.unlikePost(req, res);
  }
);

// @route POST api/posts/comment/:id
// @desc Add comment to post
// @access Private
router.post(
  "/comment/:id",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    postsController.commentPost(req, res);
  }
);

// @route DELETE api/posts/comment/:id/:comment_id
// @desc Delete comment
// @access Private
router.delete(
  "/comment/:id/:comment_id",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    postsController.deleteComment(req, res);
  }
);
module.exports = router;
