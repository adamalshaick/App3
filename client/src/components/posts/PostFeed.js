import React, { Component } from "react";
import PropTypes from "prop-types";
import PostItem from "./PostItem";

class PostFeed extends Component {
  render() {
    const { posts, profiles, profile } = this.props;
    const postFeed = posts.map(post => (
      <PostItem key={post._id} post={post} profile={profile} />
    ));
    return <div className="col-12">{postFeed}</div>;
  }
}

PostFeed.propTypes = {
  posts: PropTypes.array.isRequired
};

export default PostFeed;
