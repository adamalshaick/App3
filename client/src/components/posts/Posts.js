import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import PostForm from "./PostForm.js";
import { getPosts } from "../../actions/postActions";
import PostFeed from "./PostFeed";
import Loading from "../common/Loading.js";

class Posts extends Component {
  componentDidMount() {
    this.props.getPosts();
  }

  render() {
    const { posts, loading } = this.props.post;
    let postContent;

    if (posts === null || loading) {
      postContent = <Loading />;
    } else {
      postContent = <PostFeed posts={posts} />;
    }

    return (
      <div>
        <PostForm />
        <div className="mt-5">{postContent}</div>
      </div>
    );
  }
}

Posts.propTypes = {
  getPosts: PropTypes.func.isRequired,
  post: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  post: state.post
});

export default connect(
  mapStateToProps,
  { getPosts }
)(Posts);
