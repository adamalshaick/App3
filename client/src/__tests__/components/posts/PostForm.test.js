import React from "react";
import { PostForm } from "../../../components/posts/PostForm";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

let wrapper;

const Props = {
  errors: {}
};

wrapper = shallow(<PostForm {...Props} />);

describe("render component", () => {
  it("renders component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
