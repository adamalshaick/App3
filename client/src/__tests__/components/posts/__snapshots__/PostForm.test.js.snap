// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<div
  className="mb-3"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <TextAreaFieldGroup
        className="form-control form-control-lg"
        name="text"
        onChange={[Function]}
        placeholder="Add a post"
        value=""
      />
    </div>
    <button
      className="btn btn-outline-primary btn-sm float-right"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
